staging_upload.py

```diff
@@ -47,158 +65,152 @@ def die(msg):
-
-
-def file_sha512sum(file_path):
-    """ From https://github.com/mozilla-releng/mozapkpublisher/blob/3a2fcf264e7ac900ad9476ae5beeb2878585bcaa/mozapkpublisher/utils.py#L19
-    """
-    bs = 65536
-    hasher = hashlib.sha512()
-    with open(file_path, 'rb') as fh:
-        buf = fh.read(bs)
-        while len(buf) > 0:
-            hasher.update(buf)
-            buf = fh.read(bs)
-    return hasher.hexdigest()
-
-
-class CheckSumMismatch(Exception):
-    """ From https://github.com/mozilla-releng/mozapkpublisher/blob/3a2fcf264e7ac900ad9476ae5beeb2878585bcaa/mozapkpublisher/exceptions.py#L16
-    """
-    def __init__(self, checked_file, expected, actual):
-        super(CheckSumMismatch, self).__init__(
-            'Downloading "{}" failed!. Checksum "{}" was expected, but actually got "{}"'
-            .format(checked_file, expected, actual)
+        mar_path = abs_paths['mar']
+        check_mar_against_checksum_file(mar_path, abs_paths['checksums'])
+
+        path_with_sha512sums_per_platform[platform] = {
+            'mar': mar_path,
+            'sha512': file_sha512sum(mar_path)
+        }
+
+    log.info('Checksum verification passed')
+    return path_with_sha512sums_per_platform
+
+
+async def upload(context, path_with_sha512sums_per_platform):
+    tasks = []
+
+    for platform, details in path_with_sha512sums_per_platform.items():
+        context.release_props['platform'] = platform
+
+        for locale in ALL_LOCALES:
+            context.task['payload']['locale'] = locale
+            mapping_manifest = generate_beetmover_manifest(context)
+
+            artifact_name = 'target.complete.mar'
+            artifacts_to_beetmove = {
+                locale: {
+                    artifact_name: details['mar'],
+                },
+            }
+
+            log.info('Beetmoving {}/{}...'.format(platform, locale))
+            tasks.append(
+                asyncio.ensure_future(
+                    move_beets(context, artifacts_to_beetmove, mapping_manifest)
+                )
+            )
+
+    await raise_future_exceptions(tasks)
+    await _upload_general_sha512sums_file(context, path_with_sha512sums_per_platform)
+
+
+async def _upload_general_sha512sums_file(context, path_with_sha512sums_per_platform):
+    general_shasum_file = _generate_sha512sums_file(context, path_with_sha512sums_per_platform)
+    await upload_to_s3(
+        context=context,
+        # TODO this logic should not duplicate the one in move_beets()
+        s3_key='pub/devedition/candidates/{}-candidates/build{}/SHA512SUMS'.format(VERSION, BUILD_NUMBER),
+        path=general_shasum_file
+    )
+
+
+def _generate_sha512sums_file(context, path_with_sha512sums_per_platform):
+    # TODO this logic should not duplicate the one in move_beets()
+    lines = [
+        '{sha512}  update/{platform}/{locale}/devedition-{version}.complete.mar'.format(
+            sha512=details['sha512'], platform=platform, locale=locale, version=VERSION,
```

Comment:
Nope, `details` is defined in the list comprehension. I don't mind changing the name if it's confusing